coconut-svsm/svsm
# Add support for vTPM and its runtime protocol

This is a sketch of how the vTPM and the vTPM protocol would work in the SVSM, for now running in CPL0. 

The vTPM is initialized (manufactured) on every boot. If OVMF and the Linux guest kernel have proper tpm drivers, they should be able to communicate with the SVSM vTPM. See the section Dependencies below.

Once this is merged I can post a PR for the attestation protocol.

## Dependencies

OVMF
```
$ git clone https://github.com/coconut-svsm/edk2.git
$ cd edk2/
$ git fetch origin pull/2/head:edk2-pr2
$ git checkout edk2-pr2
$ git submodule init
$ git submodule update

$ export PYTHON3_ENABLE=TRUE
$ export PYTHON_COMMAND=python3
$ make -j16 -C BaseTools/
$ source ./edksetup.sh --reconfig
$ build -a X64 -b DEBUG -t GCC5 -D DEBUG_ON_SERIAL_PORT -D DEBUG_VERBOSE -DTPM2_ENABLE -p OvmfPkg/OvmfPkgX64.dsc
```

Guest kernel:

The linux branch below is based on the [PR#2](https://github.com/coconut-svsm/linux/pull/2) posted against the coconut-svsm/linux repository with additional two patches for the TPM platform driver.

`CONFIG_TCG_PLATFORM` should be enabled in the guest. The same kernel also needs to be installed on the host.

```
$ git clone https://github.com/cclaudio/linux.git
$ cd linux
$ git checkout svsm-host-guest-vtpm
```

QEMU:

Follow the instructions described in [INSTALL.md](https://github.com/coconut-svsm/svsm/blob/main/Documentation/INSTALL.md#building-qemu) to build the QEMU branch below.

```
$ git clone https://github.com/coconut-svsm/qemu.git
$ cd qemu
$ git fetch origin pull/2/head:qemu-pr2
$ git co qemu-pr2
```

## How to test it

Follow the [INSTALL.md](https://github.com/coconut-svsm/svsm/blob/main/Documentation/INSTALL.md#building-the-coconut-svsm) to build this PR and run a QEMU guest.

During the boot you should be able to see messages from OVMF and the linux kernel like these:

```
Loading PEIM at 0x0007F7FC000 EntryPoint=0x0007F7FD906 Tcg2ConfigPei.efi
Found SVSM TPM
Tcg2ConfigPeimEntryPoint
Tcg2ConfigPeimEntryPoint: TPM2 detected

[...]

ReadPcr - HashAlg = 0x0004, Pcr[00], digest = 51 C3 23 DE 0C 0C 69 4F 46 01 CD D0 2B EB 58 FF 13 62 9F 74 
ReadPcr - HashAlg = 0x000B, Pcr[00], digest = FC EC B5 6A CC 30 38 62 B3 0E B3 42 C4 99 0B EB 50 B5 E0 AB 89 72 24 49 C2 D9 A7 3F 37 B0 19 FE 
ReadPcr - HashAlg = 0x000C, Pcr[00], digest = 61 93 87 2D C7 23 D5 33 E3 BB 45 FB 0A EE C1 35 48 AD DE 71 11 DF 93 A4 D7 0C B1 B5 77 CE 31 10 4A C9 DF BC B8 76 BD 07 F7 7D 2C E4 B3 F7 33 DF

[...]

[    0.907842] SEV: SNP guest platform device initialized.
[    0.910991] SEV: SNP SVSM VTPM platform device initialized
[    0.993792] tpm tpm: TPM 2.0 platform device

```

From guest userspace you can also install the `tpm2-tools` package and use it to communicate with the SVSM-vTPM (e.g. `tpm2_pcrread` without arguments)
